python/pyspark/sql/dataframe.py

```diff
@@ -31,23 +31,23 @@
 
 from pyspark import copy_func, since, _NoValue
 from pyspark.rdd import RDD, _load_from_socket, _local_iterator_from_socket, \
-    ignore_unicode_prefix, PythonEvalType
-from pyspark.serializers import ArrowCollectSerializer, BatchedSerializer, PickleSerializer, \
+    ignore_unicode_prefix
+from pyspark.serializers import BatchedSerializer, PickleSerializer, \
     UTF8Deserializer
 from pyspark.storagelevel import StorageLevel
 from pyspark.traceback_utils import SCCallSiteSync
 from pyspark.sql.types import _parse_datatype_json_string
 from pyspark.sql.column import Column, _to_seq, _to_list, _to_java_column
 from pyspark.sql.readwriter import DataFrameWriter
 from pyspark.sql.streaming import DataStreamWriter
-from pyspark.sql.types import IntegralType
 from pyspark.sql.types import *
-from pyspark.util import _exception_message
+from pyspark.sql.pandas.conversion import PandasConversionMixin
+from pyspark.sql.pandas.map_ops import PandasMapOpsMixin
 
 __all__ = ["DataFrame", "DataFrameNaFunctions", "DataFrameStatFunctions"]
 
 
-class DataFrame(object):
+class DataFrame(PandasMapOpsMixin, PandasConversionMixin):
```

Comment:
Could you explain why use mixins here? Since these are never used outside this context nor (as far as I can tell) mimic the Scala counterpart, the approach seems a bit counterintuitive.

Wouldn't make more sense to use proper instances, same way as `DataFrameNaFunctions` or `DataFrameStatFunctions`?

This question also applies to `GroupedData` and its mxin.

Reply:
Ah you mean API usages like:

```python
df.pandas.mapInPandas(...)
```

via defining an instance like `DataFrameNaFunctions` or `DataFrameStatFunctions`? I actually considered this approach first. If I could write it from the first place, I think I could consider this option without concerning about backward compatibility.

If you mean defining `DataFrameNaFunctions` or `DataFrameStatFunctions` and calling directly at `DataFrame` or `SparkSession`, the API names have to be exposed at at `DataFrame` or `SparkSession` level .. which is ugly.

Also, Scala guys are pretty used to mixins. If we think self type trait, I think it fits to the purpose.

Lastly (trivial but still), I wanted to make the codes a bit more pandas friendly by mimicking in case we can have some more pandas dev guys.

Reply:
In general I am trying to get a better feeling of overall purpose of such refactoring.

As for now there is no indication that any of these mixins will be ever used outside the current context (`DataFrame` and `GroupedData`). That impression is further enforced by explicit type checks ([here](https://github.com/apache/spark/blob/cfd78393e76f454503e7cf5416f6d56f1efffd0a/python/pyspark/sql/pandas/group_ops.py#L96) and [here](https://github.com/apache/spark/blob/cfd78393e76f454503e7cf5416f6d56f1efffd0a/python/pyspark/sql/pandas/map_ops.py#L64)). So that doesn't really seem like a canonical use of mixin, especially when base core `DataFrame` is not designed for extensiblity. 

> Ah you mean API usages like:
>     
>    df.pandas.mapInPandas(...)

That's one possible approach though not the one I was thinking about. I assumed (though I am not sure, as the amount of code moved, excluding docs, message and some static stuff is negligible, and  tightly coupled with `DataFrame` anyway) that the point is maintainability.

So possible approach is either direct 

        def __init__(self, ...):
            ...
            self._pandasMapOpsMixin = PandasMapOpsMixin(self)
            ...

        def mapInPandas(self, udf):
            return self._pandasMapOpsMixin.mapInPandas(udf)

or indirect (by overwriting `__geattr__`).

Reply:
Yeah, 

```python
    def __init__(self, ...):
        ...
        self._pandasMapOpsMixin = PandasMapOpsMixin(self)
        ...

    def mapInPandas(self, udf):
        return self._pandasMapOpsMixin.mapInPandas(udf)
```

This one, the `DataFrame` or `SparkSession` have to expose the method themselves in their own classes. So, if you want to add some new pandas APIs, then it should be added into `PandasMapOpsMixin` and then also `DataFrame`.

The point I wanted to say is, the current `DataFrame` itself is closer to the Scala side's `Dataset` since it does not expose such APIs.

Also, I believe what I am doing is what self type trait is supposed to be doing. It's coupled to specific type and other types cannot implement this trait.

Reply:
I meant:


```python
class PandasMapOps(object):
    def toPandas(self, ...):
        ...
        return # pandas DataFrame

    # other Pandas <> PySpark APIs
```


```python
class DataFrame(object):
    def __init__(self, ...):
        ...
        self.pandas_map_ops = PandasMapOps(self)

    # other DataFrame APIs equivalent to Scala side.

    def toPandas(self, ...):
        return pandas_map_ops.toPandas(...)
```

vs


```python
class PandasAPIsMaxin(object):
    def toPandas(self, ...):
        ...
        return # pandas DataFrame

    # other Pandas <> PySpark APIs
```

```python
class DataFrame(PandasAPIsMaxin):
    # other DataFrame APIs equivalent to Scala side.
```

I thought the latter is better.

Reply:
I think both are fine. It's internal so we can change it later. To save the effort, let's keep what it is in this PR, i.e. the second option.

Reply:
@cloud-fan 

> I think both are fine. It's internal so we can change it later.

It is hardly internal, considering that as mixed classes are "public" (as much as it is meaningful to say about access control in Python) and, in contrast to other changes proposed here, all shifted non-static methods are part of external API. The impact so far is small though, if that's what you mean.

@HyukjinKwon 

> Also, I believe what I am doing is what self type trait is supposed to be doing. It's coupled to specific type and other types cannot implement this trait.

Such patterns, or its closes Python equivalents (see for example Django mixins for class-based views) typically indicate two things:

- Potential for inheritance, which is clearly not the case, given both Spark API and design of the `DataFrame` class.
- Non-obligatory character, which once again is not the case.

So I guess the question I am trying to ask is - "what future planned changes justify such move" - as for now it seems mostly obsolete, and less effort path, given implied time pressure, would be to keep things as-is.

Reply:
> It is hardly internal, considering that as mixed classes are "public" (as much as it is meaningful to say about access control in Python) and, in contrast to other changes proposed here, all shifted non-static methods are part of external API. The impact so far is small though, if that's what you mean.

Either way the same APIs are exposed in the same class. It's also easily switch to each other with minimised change.

Reply:
> Either way the same APIs are exposed in the same class. It's also easily switch to each other with minimised change.

While it is true, it is not equivalent from perspective of type checkers. And since a lot of related discussions circles around typing, it might be something to consider. 

Nonetheless, I can take a hint.

Reply:
What I am trying to do is basically separate pandas-related APIs, and Scala-equivalent APIs.

So, if you want to make some changes in pandas-related APIs,  you make changes in `PandasMapOpsMixin`.
If you need to make some changes in PySpark, you make changes at `DataFrame`.
Both are completely separate in the current way. Given the suggestion, it will mix up.

Reply:
+1 on @HyukjinKwon 's suggestion.